### ISSUE TITLE
Subscribe listener called on app resumed after background mode 

### ISSUE THREAD

**Issue by liquidprogrammer** [BranchMetrics/react-native-branch-deep-linking#201]
Hello!

I have an issue with my app opened from deep link on resuming from background mode (fresh open is good). If I send my app to background (switch to another app without killing my app) and then resume to it, callback passed to *branch.subscribe* will be called with initial url event. I'm sure that i'm subscribing only once.

Why it is called again? How can i disable that issue locally (may be in native code) as a fast hotfix?
Thanks!

Issue is the same on iOs and Android. I use:
**ReactNative: 0.43.2**
**react-native-branch: 2.0.0-beta.3**

**Comment by liquidprogrammer** [comment]
The callback is called by event *RNBranch.initSessionSuccess*. Seems like when app resumes from background a new session will be created (or re-init)?

**Comment by jdee** [comment]
Hi @liquidprogrammer. When this happens, what are the arguments you receive? Are the `params` null or non-null? Are you using non-Branch links (to your own non-Branch domain) or only Branch links? Can you reproduce this using the examples in this repo, e.g. `webview_example`? You can follow [these instructions](https://github.com/BranchMetrics/react-native-branch-deep-linking/tree/master/examples#automatically-set-up-ios-branch-configuration-experimental) to easily set up the examples to use your Branch keys and domains on iOS (or do it manually for either platform). I could not reproduce the problem just now using the example apps.

Also, there is now a 2.0.0-beta.4 you can try to see if that has any impact on this issue.

**Comment by liquidprogrammer** [comment]
Hi @jdee, i do not receive any arguments (don't use them for now) - just uri. My app can handle both branch links and non-branch links. It doesn't matter what link opened an app (branch or non-branch), listener will be fired on resuming from background

**Comment by liquidprogrammer** [comment]
I have just tried 2.0.0-beta.4 version - issue still persists :(

**Comment by jdee** [comment]
I could not reproduce the problem where a non-null `uri` field is received when the app enters the foreground. It is expected that you should receive `{ error: null, params: null, uri: null }` whenever the app enters the foreground, except in certain circumstances, when you might receive a deferred deep link. This is, essentially, consistent with the behavior of every other Branch SDK.

You have to use the `params` when routing a Branch link. The `uri` field is always the original link that opened the app, before being passed to Branch. For Branch links, this will be a short link, which is opaque. You need to get the `params` back from the Branch SDK/API in order to know what the link is. The `uri` field is generally only useful for non-Branch links, in which case you get null for `params`.

Unfortunately, the data model for this (RN) SDK has been a little different from all the others. I've been reluctant to change it until now, but since this is 2.0.0-beta, the time has come. See #203. This will be merged to master momentarily and released soon in beta.5. The `uri` field should be essentially considered deprecated.

See the examples, particularly webview_example, for how this is used.

- [getting data from params in subscribe callback](https://github.com/BranchMetrics/react-native-branch-deep-linking/blob/master/examples/webview_example/src/App.js#L22)
- [creating the BUO with canonicalUrl, title, contentImageUrl](https://github.com/BranchMetrics/react-native-branch-deep-linking/blob/master/examples/webview_example/src/Article.js#L34)
- [calling showShareSheet with link parameters and control parameters (also returned via params in subscribe callback)](https://github.com/BranchMetrics/react-native-branch-deep-linking/blob/master/examples/webview_example/src/Article.js#L70)

If you are able to test the master branch and verify that it fixes your problem, I can release it soon as beta.5.

**Comment by liquidprogrammer** [comment]
Hi, @jdee. I have just tried the master branch and seems like the problem gone away! Now i don't use *uri* property and work with *+non_branch_link* property inside *params* object, and everything seems to be okay :)

Please, let me know when *beta.5* will be released and after that, i think, we can close that issue. Thank you for your help!

**Comment by jdee** [comment]
@liquidprogrammer Release 2.0.0-beta.5 is available in NPM now with this change. I'm closing this issue. Please get in touch if you have any further problems.

**Comment by liquidprogrammer** [comment]
Hi, @jdee.
Everything is okay now on iOs, but on Android issue still persists:
1. I launch my app with a deep link
2. Send it to background (not kill)
3. Open app with deep link again
4. Now switching between apps (without clicking any links) will trigger subscribe event with **+non_branch_link** property and it won't be null. On iOs **params** object doesn't have that property in **params** on app resumed

Am i missing something?
Thanks!

**Comment by jdee** [comment]
I can reproduce this issue. The `+non_branch_link` is coming from the Branch Android SDK, however. It's possible that it's being caused by how it's being used in the react-native-branch library, but it might also be a bug in the Android SDK or elsewhere upstream.

For now, I noticed that I'm getting a Branch link in the `+non_branch_link` field. Is that consistent with your results? If so, you can temporarily filter those events out in case the `+non_branch_link` contains an app.link URL or one to your custom Branch domain.

```
05-23 16:20:28.687 18518-19375/com.webview_example I/ReactNativeJS: Branch params: {"+non_branch_link":"https://bnctestbed.test-app.link/piXQDmekgD","+clicked_branch_link":false,"+is_first_session":false}
```

**Comment by liquidprogrammer** [comment]
Hi, @jdee.
It also happens with non-branch links. For now i have a workaround for that like this (ignore the last link):

```javascript
let lastUri = ''

branch.subscribe(event => {
  ...
  let uri = event.params['+non_branch_link']
  if (lastUri === uri) {
    return
  }
  
  lastUri = uri
  // new link, do smth here
})
```

It is a hack, but helps me to avoid processing the same link again and again :)
Should we reopen that issue or create a new one?

**Comment by jdee** [comment]
@liquidprogrammer I've opened #216 to track this issue. The `+non_branch_link` is returned by the native SDK, but it can't be reproduced using the native SDK alone, so it's probably something in the react-native-branch codebase. I'm trying to dig it out now.

**Comment by liquidprogrammer** [comment]
@jdee Great, thank you!
